sql-tuning-best-practice.md

```diff
@@ -0,0 +1,714 @@
+3. After confirming that it is not a Point Get, the AST goes to the Logical Transformation, which rewrites the SQL logically based on certain rules.
+4. The AST that has gone through Logical Transformation will undergo Cost-Based Optimization.
+5. During Cost-Based Optimization, the optimizer considers statistics to determine how to select specific operators and finally generates a physically executable physical execution plan.
+6. The generated physical execution plan is sent to the SQL Executor of the TiDB database for execution.
+7. Unlike traditional databases, TiDB databases need to push down the execution plan to different TiKV for reading and writing data.
+
+![workflow](/media/sql-tuning/workflow.png)
+
+## Optimizer Fundamentals
+
+TiDB uses a cost-based optimizer (CBO) to determine the most efficient execution plan for a SQL statement. This optimizer evaluates different execution strategies and chooses the one with the lowest estimated cost. The cost is influenced by factors such as:
+
+- SQL
+- Schema Design
+- Statistics
+  - Table
+  - Index
+  - Column
+
+Based on the input, The cost model will produce the execution plan, which includes the details how the system execute the sql, including 
+- Access Method
+- Join Method
+- Join Order
+
+The optimizer is as good as the information it receives. Therefore, ensuring up-to-date statistics and well-designed indexes is critical.
+
+## Statistics Management
+
+Statistics are essential to the TiDB optimizer. TiDB uses statistics as input to the optimizer to estimate the number of rows processed in each plan step for a SQL statement.
+
+statistics is generally divided into two levels: table level and column level. 
+- For table-level statistics, it includes the total number of rows in the table and the number of rows that have been modified since the last collection of statistics. 
+- The column-level statistics information is more abundant, including histograms, Count-Min Sketch, Top-N (values or indexes with the highest occurrences), distribution and quantity of different values, and the number of null values, and so on.
+
+To ensure the statistics are healthy and representative, you can use the following commands:
+
+1. `SHOW STATS_META`: This command provides metadata about table statistics.
+2. `SHOW STATS_HEALTHY`: This command shows the health status of table statistics.
+
+For example, you can use:
+
+```SQL
+tidb> SHOW STATS_META WHERE table_name='T2'\G;
+*************************** 1. row ***************************
+ Db_name: test
+ Table_name: T2
+Partition_name:
+ Update_time: 2023-05-11 02:16:50
+ Modify_count: 20000
+ Row_count: 20000
+1 row in set (0.03 sec)
+
+tidb> SHOW STATS_HEALTHY WHERE table_name='T2'\G;
+*************************** 1. row ***************************
+    Db_name: test
+  Table_name: T2
+Partition_name:
+    Healthy: 0
+1 row in set (0.00 sec)
+
+```
```

Comment:
````suggestion
```sql
SHOW STATS_META WHERE table_name='T2'\G;
```

```
*************************** 1. row ***************************
 Db_name: test
 Table_name: T2
Partition_name:
 Update_time: 2023-05-11 02:16:50
 Modify_count: 20000
 Row_count: 20000
1 row in set (0.03 sec)
```

```sql
SHOW STATS_HEALTHY WHERE table_name='T2'\G;
```

```
*************************** 1. row ***************************
    Db_name: test
  Table_name: T2
Partition_name:
    Healthy: 0
1 row in set (0.00 sec)
```
````

The output isn't valid SQL